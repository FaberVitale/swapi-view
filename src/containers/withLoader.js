//flow
import * as React from "react";
import { getDisplayName, identity } from "../util/functions";
import type { Match } from "react-router-dom";
/* Higher order component that given the props passed renders 3 different components:
 *  
 * Props: 
 *  data:  ?D if null or undefined data is not loaded
 *  error: ?E if null or undefined error has not occurred
 *  load:  () => void function that is called when data is not null or undefined
 *  search: react-router's location.search string
 *  match: React-router's Match, used to call load on route changes.
 *  
 * Components:
 * 1 - If props.error is NOT null or undefined renders 
 *    <ErrorComponent error={error}{...otherProp} />.
 * 2 - If props.data is NOT null or undefined renders 
 *    <WrappedComponent data={data} {...otherProp} />.
 * 3 - Otherwise renders <LoadingComponent {...otherProp} />.
 * 
 * It will call load again onDidUpdate if it is in Loading state and route has changed
 * 
 *  Options:
 *    mapProps: equivalent to Recompose mapProps it is applied only to the wrappedComponent
 *      description: a transformation of the props passed to the WrappedComponent, props.data included.
 *      default: obj => obj
 *    LoadingComponent: React.ElementType (a React Component or a string).
 *      description: Component rendered if props.error and props.data are both null or undefined
 *      default:   () => <p>Loading...</p>
 *    ErrorComponent: React.ElementType (a React Component or a string).
 *      description: Component rendered if props.error is not null or undefined
 *      it passed the load function renamed retry, error, and other props
 *       default: () => <p>An error has occurred.</p>.
*/

type Options = {|
  mapProps: (props: {}) => {},
  LoadingComponent: React.ComponentType<any>,
  ErrorComponent: React.ComponentType<any>
|};

type Props<D, E> = {
  match: Match,
  search: string,
  data?: ?D,
  error?: ?E,
  load: () => void
};

type HocType = <D, E>(
  opt?: Options
) => (React.ComponentType<any>) => React.Component<Props<D, E>>;

const defaultOpt: Options = {
  mapProps: identity,
  LoadingComponent: () => <p>Loading...</p>,
  ErrorComponent: () => <p>An error has occurred.</p>
};

const withLoader: HocType = (options: Options = defaultOpt) => (
  WrappedComponent: React.ComponentType<any>
) => {
  let opt: Options =
    options === defaultOpt ? options : Object.assign({}, defaultOpt, options);

  class Loader extends React.Component<Props<D, E>> {
    componentDidMount() {
      //true if undefined or null
      if (this.props.data == null && this.props.error == null) {
        this.props.load();
      }
      this._route = this.props.match.url;
      this._search = this.props.search;
    }

    componentDidUpdate() {
      if (
        (this.props.match.url !== this._route ||
          this._search !== this.props.search) &&
        this.props.data == null &&
        this.props.error == null
      ) {
        this.props.load();

        this._search = this.props.search;
        this._route = this.props.match.url;
      }
    }
    render() {
      const { data, error, load, ...rest } = this.props;

      let nextProps, Comp;

      if (error != null) {
        nextProps = Object.assign(rest, { error, retry: load });
        Comp = opt.ErrorComponent;
      } else if (data != null) {
        nextProps = opt.mapProps({ data, ...rest });
        Comp = WrappedComponent;
      } else {
        nextProps = rest;
        Comp = opt.LoadingComponent;
      }

      return React.createElement(Comp, nextProps);
    }
  }

  if (process.env.NODE_ENV !== "production") {
    Loader.displayName = getDisplayName("Loader", WrappedComponent);
  }
  return Loader;
};

export default withLoader;
