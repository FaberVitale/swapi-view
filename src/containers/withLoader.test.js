import React from "react";
import withLoader from "./withLoader";
import { MemoryRouter, Route } from "react-router-dom";
import { shallow, mount } from "enzyme";

describe("containers/withRouter", () => {
  const Comp = ({ data }) => <p>{JSON.stringify(data)}</p>;
  const load = jest.fn(() => {});

  let push;
  let wrapper;

  const GetPush = props => {
    push = props.history.push;
    return null;
  };
  const NoOpt = withLoader()(Comp);

  const renderNoOptFunc = rest => ({ match, location, history }) => [
    <GetPush key={0} history={history} />,
    <NoOpt key={1} {...rest} match={match} search={location.search} />
  ];

  const TestNoOptComp = ({ children, ...rest }) => (
    <MemoryRouter initialEntries={["/"]} initialIndex={0}>
      <Route exact path="*" render={renderNoOptFunc(rest)} />
    </MemoryRouter>
  );

  const baseProps = {
    match: { url: "/" },
    search: ""
  };

  const loadingProps = {
    data: null,
    error: null,
    load
  };

  const errorProps = {
    data: null,
    error: "err",
    load
  };

  const dataProps = {
    data: { a: 2, b: 3 },
    error: null,
    load
  };

  afterEach(() => {
    load.mockClear();
    wrapper.unmount();
  });

  it("default opt, match the snapshots", () => {
    wrapper = shallow(<TestNoOptComp {...loadingProps} />);

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.setProps(errorProps).html()).toMatchSnapshot();
    expect(wrapper.setProps(dataProps).html()).toMatchSnapshot();
  });

  it("calls load if error == null && data == null on DidMount hook", () => {
    wrapper = mount(<TestNoOptComp {...loadingProps} />);
    expect(load.mock.calls).toHaveLength(1);
  });

  it("on match.url or search change if loading, calls load again", () => {
    wrapper = mount(<TestNoOptComp {...loadingProps} />);

    expect(load.mock.calls).toHaveLength(1);

    push("/a");

    expect(load.mock.calls).toHaveLength(2);

    push("/a?q=a");

    expect(load.mock.calls).toHaveLength(3);
  });

  it("if error != null, it passes down load as retry", () => {
    wrapper = mount(<TestNoOptComp {...errorProps} />);

    expect(
      wrapper
        .find("p")
        .parent()
        .prop("retry")
    ).toBe(load);
  });

  it("uses mapProps if defined and  data != null && error == null", () => {
    let mapProps = jest.fn(props => ({ data: 3 }));

    const MapProps = withLoader({
      mapProps
    })(Comp);

    const TestMapPropsComp = ({ children, ...rest }) => (
      <MapProps {...baseProps} {...rest} />
    );

    wrapper = mount(<TestMapPropsComp {...loadingProps} />);

    expect(mapProps.mock.calls).toHaveLength(0);

    wrapper.setProps(errorProps);

    expect(mapProps.mock.calls).toHaveLength(0);

    wrapper.setProps(dataProps);

    expect(mapProps.mock.calls).toHaveLength(1);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("uses ErrorComponent and LoadingComponent, if defined", () => {
    const errorMsg = "error";
    const loadingMsg = "loading...";

    const OptComp = withLoader({
      ErrorComponent: () => <p>{errorMsg}</p>,
      LoadingComponent: () => <p>{loadingMsg}</p>
    })(Comp);

    wrapper = mount(<OptComp {...baseProps} {...loadingProps} />);

    expect(wrapper.find("p").text()).toBe(loadingMsg);

    wrapper.setProps(Object.assign({}, baseProps, errorProps));

    expect(wrapper.find("p").text()).toBe(errorMsg);

    wrapper.setProps(Object.assign({}, baseProps, dataProps));

    const dataText = wrapper.find("p").text();

    expect(dataText).not.toBe(errorMsg);
    expect(dataText).not.toBe(loadingMsg);
  });
});
