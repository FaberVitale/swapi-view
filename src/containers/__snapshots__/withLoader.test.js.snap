// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/withRouter default opt, match the snapshots 1`] = `"<p>Loading...</p>"`;

exports[`containers/withRouter default opt, match the snapshots 2`] = `"<p>An error has occurred.</p>"`;

exports[`containers/withRouter default opt, match the snapshots 3`] = `"<p>{&quot;a&quot;:2,&quot;b&quot;:3}</p>"`;

exports[`containers/withRouter uses mapProps if defined and  data != null && error == null 1`] = `"<p>3</p>"`;
